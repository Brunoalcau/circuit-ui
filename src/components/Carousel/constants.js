export const ASPECT_RATIO = 1.8;

export const ANIMATION_DURATION = 640;

export const SLIDE_DURATION = 5000;

export const SLIDE_DIRECTIONS = {
  FORWARD: 'forward',
  BACK: 'back'
};
