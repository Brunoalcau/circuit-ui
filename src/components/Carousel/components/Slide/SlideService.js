import { SLIDE_DIRECTIONS } from '../../constants';

export function getStackOrder(slideIndex, step, prevStep, slideDirection) {
  const isActive = step === slideIndex;
  const wasActive = prevStep === slideIndex;

  if (slideDirection === SLIDE_DIRECTIONS.FORWARD && isActive) {
    return 2;
  }

  if (slideDirection === SLIDE_DIRECTIONS.FORWARD && wasActive) {
    return 1;
  }

  if (slideDirection === SLIDE_DIRECTIONS.BACK && isActive) {
    return 1;
  }

  if (slideDirection === SLIDE_DIRECTIONS.BACK && wasActive) {
    return 2;
  }

  return slideIndex ? -slideIndex : 0;
}

export function shouldAnimate(slideIndex, step, prevStep, slideDirection) {
  const isActive = step === slideIndex;
  const wasActive = prevStep === slideIndex;

  if (slideDirection === SLIDE_DIRECTIONS.FORWARD) {
    return isActive;
  }

  if (slideDirection === SLIDE_DIRECTIONS.BACK) {
    return wasActive;
  }

  return false;
}

export function getDynamicWidth(width) {
  return width ? `${width}px` : '100%';
}
