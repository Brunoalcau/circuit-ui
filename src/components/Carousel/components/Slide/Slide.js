import React from 'react';
import PropTypes from 'prop-types';
import styled, { css, keyframes } from 'react-emotion';

import { childrenPropType } from '../../../../util/shared-prop-types';
import * as SlideService from './SlideService';
import { ANIMATION_DURATION, SLIDE_DIRECTIONS } from '../../constants';

const baseStyles = ({ index, stackOrder, width }) => css`
  label: carousel__slide;
  width: 100%;
  flex-grow: 0;
  flex-shrink: 0;
  flex-basis: ${width};
  transform: translate3d(${-index * 100}%, 0, 0);
  backface-visibility: hidden;
  position: relative;
  z-index: ${stackOrder};
`;
const Wrapper = styled('div')(baseStyles);

const slideIn = keyframes`
  from {
    width: 0%;
  }
  to {
    width: 100%;
  }
`;
const slideOut = keyframes`
  from {
    width: 100%;
  }
  to {
    width: 0%;
  }
`;
const animationStyles = ({
  theme,
  isAnimating,
  animationDuration,
  animationName
}) => css`
  box-shadow: 0 0 1px rgba(0, 0, 0, 0.05);
  overflow: hidden;
  will-change: width;
  ${isAnimating &&
    css`
      animation-name: ${animationName};
      animation-duration: ${animationDuration}ms;
      animation-fill-mode: forwards;
      animation-timing-function: ${theme.transitions.easeInOutCubic};
    `};
`;
const Inner = styled('div')(animationStyles);

const dynamicWidthStyles = ({ width }) => css`
  width: ${width};
`;
const Content = styled('div')(dynamicWidthStyles);

const Slide = ({
  index,
  step,
  prevStep,
  slideSize,
  slideDirection,
  animationDuration,
  children,
  ...props
}) => {
  const stackOrder = SlideService.getStackOrder(
    index,
    step,
    prevStep,
    slideDirection
  );
  const shouldAnimate = SlideService.shouldAnimate(
    index,
    step,
    prevStep,
    slideDirection
  );
  const dynamicWidth = SlideService.getDynamicWidth(slideSize.width);
  const animationName =
    slideDirection === SLIDE_DIRECTIONS.FORWARD ? slideIn : slideOut;

  return (
    <Wrapper
      index={index}
      stackOrder={stackOrder}
      width={dynamicWidth}
      {...props}
    >
      <Inner
        isAnimating={slideSize.width && shouldAnimate}
        animationName={animationName}
        animationDuration={animationDuration}
      >
        <Content width={dynamicWidth}>{children}</Content>
      </Inner>
    </Wrapper>
  );
};

Slide.propTypes = {
  /**
   *  Index of a slide in a carousel (required for animation)
   */
  index: PropTypes.number,
  /**
   * Current active index of a carousel (required for animation)
   */
  step: PropTypes.number,
  /**
   * Previous active index of a carousel (required for animation)
   */
  prevStep: PropTypes.number,
  /**
   * Dimensions of a carousel (width of the slide is required for animation)
   */
  slideSize: PropTypes.shape({
    width: PropTypes.number,
    height: PropTypes.number
  }),
  /**
   * Indicates slide direction of a carousel (required for animation)
   */
  slideDirection: PropTypes.oneOf([
    SLIDE_DIRECTIONS.FORWARD,
    SLIDE_DIRECTIONS.BACK
  ]),
  /**
   * Duration of animation between slides in milliseconds
   */
  animationDuration: PropTypes.number,
  /**
   * Content of a slide
   */
  children: childrenPropType
};

Slide.defaultProps = {
  index: 0,
  step: 0,
  slideSize: {},
  animationDuration: ANIMATION_DURATION
};

export default Slide;
