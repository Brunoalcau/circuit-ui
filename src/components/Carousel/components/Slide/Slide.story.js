import React from 'react';
import { storiesOf } from '@storybook/react';
import { css } from 'react-emotion';

import { GROUPS } from '../../../../../.storybook/hierarchySeparators';
import withTests from '../../../../util/withTests';

import Heading from '../../../Heading';
import Image from '../../../Image';
import Slide from './Slide';

storiesOf(`${GROUPS.COMPONENTS}|Carousel/Slide`, module)
  .addDecorator(withTests('Carousel/Slide'))
  .add('Slide with image', () => (
    <div style={{ width: '60vw' }}>
      <Slide>
        <Image src="http://www.placepuppy.net/800/500" />
      </Slide>
    </div>
  ))
  .add('Slide with text and image', () => (
    <div style={{ width: '60vw' }}>
      <Slide>
        <Image src="http://www.placepuppy.net/800/500" />
        <Heading
          css={css`
            color: #fff;
            width: 50%;
            position: absolute;
            bottom: 0;
            left: 25px;
            z-index: 2;
          `}
        >
          Get The SumUp Card Reader Today!
        </Heading>
      </Slide>
    </div>
  ));
