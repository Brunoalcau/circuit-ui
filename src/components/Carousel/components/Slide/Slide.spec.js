import React from 'react';

import Slide from './Slide';
import { SLIDE_DIRECTIONS } from '../../constants';

describe('Slide', () => {
  describe('styles', () => {
    it('should render with default styles', () => {
      const actual = create(<Slide />);

      expect(actual).toMatchSnapshot();
    });

    it('should render with forward animation styles', () => {
      const actual = create(
        <Slide
          index={0}
          step={0}
          slideSize={{ width: 800 }}
          slideDirection={SLIDE_DIRECTIONS.FORWARD}
        />
      );

      expect(actual).toMatchSnapshot();
    });

    it('should render with backward animation styles', () => {
      const actual = create(
        <Slide
          index={0}
          prevStep={0}
          slideSize={{ width: 800 }}
          slideDirection={SLIDE_DIRECTIONS.BACK}
        />
      );

      expect(actual).toMatchSnapshot();
    });
  });

  describe('accessibility', () => {
    it('should meet accessibility guidelines', async () => {
      const wrapper = renderToHtml(<Slide />);
      const actual = await axe(wrapper);

      expect(actual).toHaveNoViolations();
    });
  });
});
