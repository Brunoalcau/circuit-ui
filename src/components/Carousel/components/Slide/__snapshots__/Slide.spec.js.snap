// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slide styles should render with backward animation styles 1`] = `
.circuit-0 {
  width: 800px;
}

.circuit-4 {
  width: 100%;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 800px;
  -ms-flex-preferred-size: 800px;
  flex-basis: 800px;
  -webkit-transform: translate3d(0%,0,0);
  -ms-transform: translate3d(0%,0,0);
  transform: translate3d(0%,0,0);
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  position: relative;
  z-index: 1;
}

.circuit-2 {
  box-shadow: 0 0 1px rgba(0,0,0,0.05);
  overflow: hidden;
  will-change: width;
  -webkit-animation-name: animation-1sj1vig-slideOut;
  animation-name: animation-1sj1vig-slideOut;
  -webkit-animation-duration: 640ms;
  animation-duration: 640ms;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
}

<div
  className="circuit-4 circuit-5"
  width="800px"
>
  <div
    className="circuit-2 circuit-3"
  >
    <div
      className="circuit-0 circuit-1"
      width="800px"
    />
  </div>
</div>
`;

exports[`Slide styles should render with default styles 1`] = `
.circuit-4 {
  width: 100%;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  -webkit-transform: translate3d(0%,0,0);
  -ms-transform: translate3d(0%,0,0);
  transform: translate3d(0%,0,0);
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  position: relative;
  z-index: 0;
}

.circuit-2 {
  box-shadow: 0 0 1px rgba(0,0,0,0.05);
  overflow: hidden;
  will-change: width;
}

.circuit-0 {
  width: 100%;
}

<div
  className="circuit-4 circuit-5"
  width="100%"
>
  <div
    className="circuit-2 circuit-3"
  >
    <div
      className="circuit-0 circuit-1"
      width="100%"
    />
  </div>
</div>
`;

exports[`Slide styles should render with forward animation styles 1`] = `
.circuit-4 {
  width: 100%;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 800px;
  -ms-flex-preferred-size: 800px;
  flex-basis: 800px;
  -webkit-transform: translate3d(0%,0,0);
  -ms-transform: translate3d(0%,0,0);
  transform: translate3d(0%,0,0);
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  position: relative;
  z-index: 2;
}

.circuit-2 {
  box-shadow: 0 0 1px rgba(0,0,0,0.05);
  overflow: hidden;
  will-change: width;
  -webkit-animation-name: animation-j1aola-slideIn;
  animation-name: animation-j1aola-slideIn;
  -webkit-animation-duration: 640ms;
  animation-duration: 640ms;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
}

.circuit-0 {
  width: 800px;
}

<div
  className="circuit-4 circuit-5"
  width="800px"
>
  <div
    className="circuit-2 circuit-3"
  >
    <div
      className="circuit-0 circuit-1"
      width="800px"
    />
  </div>
</div>
`;
